### ISSUE TITLE
fix test errors with XML::Simple

### ISSUE THREAD

**Issue by mmitch** [mmitch/fritz#7]
For some time there have been some errors during the unit tests related to XML::Simple, for example for version 0.0.5 we currently have 797 build logged with passed and 4 logged with errors (see http://matrix.cpantesters.org/?dist=Net-Fritz%20v0.0.5).
The errors looked like this:
```
t/02.fritz-error.t ....... ok
Not a CODE reference at /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.22.1/5ea4/lib/site_perl/5.22.1/x86_64-linux-thread-multi-ld/XML/SAX/Expat.pm line 70.
XML::Simple called at /tmp/loop_over_bdir-1582-EHochc/Net-Fritz-v0.0.4-0/blib/lib/Net/Fritz/Box.pm line 82.
# Looks like your test exited with 255 just after 5.
t/03.fritz-box.t ......... 
Dubious, test returned 255 (wstat 65280, 0xff00)
```
I did not understand the error and could not reproduce it.

Just now there are some fresh new errors in that actually contain a helpful error message:
```
t/02.fritz-error.t ....... ok
Nonblocking reads only make sense if you're giving me file handles, y'know (well, actually sockets). Use parse_file at /home/sand/src/perl/repoperls/installed-perls/host/k93msid/perl-5.8.5/89ad/lib/site_perl/5.8.5/XML/Simple.pm line 396.
XML::Simple called at /tmp/loop_over_bdir-24803-J2UscX/Net-Fritz-v0.0.5-0/blib/lib/Net/Fritz/Box.pm line 80.
# Looks like your test exited with 255 just after 5.
t/03.fritz-box.t ......... 
Dubious, test returned 255 (wstat 65280, 0xff00)
```
source: http://www.cpantesters.org/cpan/report/940f1ada-4d22-11e7-8e8a-3af4b41257c2

The new message seems to come from an update of `XML::Simple` from 2.22 to 2.24

That is something we can act upon!

**Comment by mmitch** [comment]
v0.0.6 has been released
wait some weeks and see what the CPAN build reports say before closing this bug

**Comment by mmitch** [comment]
v0.0.6 did not fix it, here is already one failed test report: http://www.cpantesters.org/cpan/report/5a5fd130-4fb6-11e7-ab79-8a5bd5ba6f85
The error is the same as before:
```
t/02.fritz-error.t ....... ok
Nonblocking reads only make sense if you're giving me file handles, y'know (well, actually sockets). Use parse_file at /home/sand/src/perl/repoperls/installed-perls/host/k93msid/v5.26.0/d2d1/lib/site_perl/5.26.0/XML/Simple.pm line 396.
XML::Simple called at /tmp/loop_over_bdir-14577-NCML0A/Net-Fritz-v0.0.6-0/blib/lib/Net/Fritz/Box.pm line 82.
# Looks like your test exited with 255 just after 5.
```
The message about the nonblocking reads does not come from XML::Simple, but from further down the stack. I stumbled across [this bug](https://rt.cpan.org/Public/Bug/Display.html?id=73574) pointing to XML::SAX::ExpatNB

XML::Simple can use different XML parsers as backend, perhaps the error lies in there.  Currently the backend is chosen automatically.
I will try different backends to see if I can reproduce the error with one of them.
If I can break the code in this way, the fix should be to simply forse the use of a working parser.

**Comment by mmitch** [comment]
Still can't reproduce in any way, but let's try setting the parser to XML::Parser.

**Comment by mmitch** [comment]
v0.0.7 with XML::Simple set to use XML::Parser has been release
Waiting for the cpantesters build servers to pick it up.

**Comment by mmitch** [comment]
v0.0.7 is out for nearly four days: 202 tests PASS, not a single one FAILs (v0.0.6 failed twice in the first 24 hours).

The FAILed test configurations from v0.0.6 PASS in v0.0.7:
* http://matrix.cpantesters.org/?dist=Net-Fritz%20v0.0.6;os=linux;perl=5.26.0;reports=1
* http://matrix.cpantesters.org/?dist=Net-Fritz%20v0.0.7;os=linux;perl=5.26.0;reports=1

So I consider this fixed.

**Comment by kjetilk** [comment]
Hi! 

I just stumbled upon this issue and that you had stumbled upon my issue from way back ;-) I just wondered if you gained any deeper understanding of the issue, than it just being a problem with the backend parser?

Cheers,

Kjetil

**Comment by mmitch** [comment]
Sadly, I don't really know what I did there…
I was never able to reproduce it locally, so I could not debug it properly. I only had the fail messages from the CPAN build/test hosts.

As I found out about the dynamic loading of the real XML parser, I thought "well, anything can happen, I don't really know what gets installed on a test host in which sequence and perhaps sometimes a strange quirky module gets selected as the default parser".
I then tried to guess a sane default and apparently got it right on the second try ;-)
